implementations/rust/ockam/ockam_entity/src/credential/traits.rs

```diff
@@ -89,3 +100,151 @@ pub trait CredentialVerifier {
+pub struct CredentialIdentifier(String);
+
+impl Distribution<CredentialIdentifier> for Standard {
+    fn sample<R: Rng + ?Sized>(&self, rng: &mut R) -> CredentialIdentifier {
+        let id: [u8; 16] = rng.gen();
+        CredentialIdentifier(hex::encode(id))
+    }
+}
+
+impl CredentialIdentifier {
+    pub const PREFIX: &'static str = "C";
+}
+
+impl Display for CredentialIdentifier {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        let str: String = self.clone().into();
+        write!(f, "{}", &str)
+    }
+}
+
+impl Into<String> for CredentialIdentifier {
+    fn into(self) -> String {
+        format!("{}{}", Self::PREFIX, &self.0)
+    }
+}
+
+impl TryFrom<&str> for CredentialIdentifier {
+    type Error = Error;
+
+    fn try_from(value: &str) -> Result<Self> {
+        if let Some(str) = value.strip_prefix(Self::PREFIX) {
+            Ok(Self(str.into()))
+        } else {
+            Err(EntityError::InvalidProfileId.into())
+        }
+    }
+}
+
+impl TryFrom<String> for CredentialIdentifier {
+    type Error = Error;
+
+    fn try_from(value: String) -> Result<Self> {
+        Self::try_from(value.as_str())
+    }
+}
+
+#[derive(Clone, Eq, PartialEq, Debug, Serialize, Deserialize)]
+pub struct Credential {
+    id: CredentialIdentifier,
+    issuer_id: EntityIdentifier,
+    type_id: String,
+}
+
+impl Credential {
+    pub fn new(id: CredentialIdentifier, issuer_id: EntityIdentifier, type_id: String) -> Self {
+        Credential {
+            id,
+            issuer_id,
+            type_id,
+        }
+    }
+}
+
+impl Credential {
+    pub fn id(&self) -> &CredentialIdentifier {
+        &self.id
+    }
+    pub fn issuer_id(&self) -> &EntityIdentifier {
+        &self.issuer_id
+    }
+    pub fn type_id(&self) -> &str {
+        &self.type_id
+    }
+}
+
+#[derive(Clone, Serialize, Deserialize)]
+pub struct EntityCredential {
+    credential: Credential,
+    bbs_credential: BbsCredential,
+    #[serde(with = "BigArray")]
+    issuer_pubkey: [u8; 96],
+    schema: CredentialSchema,
+}
+
+impl EntityCredential {
+    pub fn credential(&self) -> &Credential {
+        &self.credential
+    }
+    pub fn bbs_credential(&self) -> &BbsCredential {
+        &self.bbs_credential
+    }
+    pub fn issuer_pubkey(&self) -> [u8; 96] {
+        self.issuer_pubkey
+    }
+    pub fn schema(&self) -> &CredentialSchema {
+        &self.schema
+    }
+}
+
+impl EntityCredential {
+    pub fn new(
+        credential: Credential,
+        bbs_credential: BbsCredential,
+        issuer_pubkey: [u8; 96],
+        schema: CredentialSchema,
+    ) -> Self {
+        EntityCredential {
+            credential,
+            bbs_credential,
+            issuer_pubkey,
+            schema,
+        }
+    }
+}
+
+pub trait CredentialProtocol {
+    fn create_credential_issuance_listener(
+        &mut self,
+        address: impl Into<Address> + Send,
```

Comment:
We have other traits that use the generic form `<A: Into<Address> + Send>` on the function - Is there any idiomatic reasoning to using `impl` for these instead? Just curious. I think we should choose one way and stick with it, if possible.

Reply:
I think impl version is more elegant, especially when you have multiple generics and when you have 2 arguments of the same type (e.g. Into<Address>, but they cannot share the same generic because one of them may be String and the other &str.
The only constraint is that you cannot use impl in struct definition

Reply:
I read a bit more and I agree it's more elegant, I will switch Identity trait to this style